### PR TITLE
fix scroll behaviour + [wip] cypress tests for scroll behaviour

### LINKED ISSUE #7454: [v2] Navigating to previously visited pages with <Link> retains scroll position
## Description

Clicking on a `<Link>` in v2 (111) appears to scroll you to a previous scroll position (if it exists) rather than scroll you to the top.

### Steps to reproduce

I've been able to recreate this in `gatsby-starter-default`:
https://gatsby-v2-beta-111-scroll.netlify.com/

(The only change is artificial padding added to the first page above the `/page-2` `<Link>`)

- Visit https://gatsby-v2-beta-111-scroll.netlify.com/
- On the home page, scroll down to the `page-2` link and click it 
- On `page-2`, now click 'Go back to the homepage'

### Expected result

You are taken back to the top of the home page

### Actual result

You are taken back to the home page but scrolled down the page (at the exact same position you were when you linked the `page-2` link)

### Environment

```
System:
    OS: macOS High Sierra 10.13.6
    CPU: x64 Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz
    Shell: 5.3 - /bin/zsh
  Binaries:
    Node: 8.11.1 - ~/.nvm/versions/node/v8.11.1/bin/node
    Yarn: 1.9.4 - ~/.nvm/versions/node/v8.11.1/bin/yarn
    npm: 5.6.0 - ~/.nvm/versions/node/v8.11.1/bin/npm
  Browsers:
    Chrome: 68.0.3440.106
    Firefox: 61.0.1
    Safari: 11.1.2
  npmPackages:
    gatsby: next => 2.0.0-beta.111
    gatsby-plugin-manifest: next => 2.0.2-beta.6
    gatsby-plugin-offline: next => 2.0.0-beta.9
    gatsby-plugin-react-helmet: next => 3.0.0-beta.4
  npmGlobalPackages:
    gatsby-cli: 1.1.58
```

### Other notes

- This behaviour doesn't occur on the v1 starter. 
- I understand that there was a switch to `reach-router`, could this be linked?
- Manually adding `window.scrollTo(0, 0) / window.scroll({top: 0})` in page `componentDidMount` had no effect, though it would work after wrapping it in an arbitrary `setTimeout` - though this yields an initial flash of content in the wrong scroll position.

By default, I imagine the desired behaviour would be for all clicked `<Link>` elements to scroll you to the top of that target page and that only using the browser back / forward buttons would retain scroll position state in this manner.

### PR DESCRIPTION
**Pull request by pieh**
fixes #7454 (again) but adding tests to catch regressions in future

problem was temp hack for reach/router was disabled by recent changes because of order of calls changes - it's restored now

### CODE REVIEW

**Comment by pieh** [comment]
Argh, of course my cypress tests locally but not in CI :)

**Comment by DSchau** [comment]
Oh right _duh_. So Travis tests are going to fail, because I broke that with the CircleCI stuff. I'll fix that in a bit. It's weird, it looks like this PR somehow used the old CircleCI setup too 🙃 

I'll take a closer look in a bit! This looks good though!

**Comment by DSchau** [comment]
@pieh see #8361 and #8362 for some fixes for this stuff! It's possible your tests are failing in CI for an unrelated reason, but should at least be closer to passing 🙃

**Comment by pieh** [comment]
> it looks like this PR somehow used the old CircleCI setup too 🙃

Hmm, I'll merge master and see if that fixes circleCI test then

**Comment by DSchau** [comment]
@pieh they passed!! Except it's a false positive, haha. Basically the way it works is that the current commit's files are what trigger (or don't trigger) the e2e tests.

I think/thought that was a good idea, but now we won't really know whether they're passing for real or not unless a change to packages/ or integration-tests/ comes through in a commit. I can also change that behavior, but the thought was that for each commit we should run only what's needed. i.e. one commit might change packages -> run e2e tests. Next commit might be a README.md change -> don't run e2e tests, etc.

**Comment by pieh** [comment]
Oh, so it doesn't check file changes in PR - just file changes in latest commit?

**Comment by DSchau** [comment]
@pieh yup. But like I said, I can change that, open to other suggestions :)

**Comment by KyleAMathews** [comment]
That makes sense as a default. This is an edge case. Is there a way to manually trigger tests to run for real?

**Comment by DSchau** [comment]
@KyleAMathews I _think_ I can set up the e2e test suite as an [approval workflow](https://circleci.com/docs/2.0/workflows/#holding-a-workflow-for-a-manual-approval). Those tests wouldn't make the same check as to whether certain files changed.

It'd be a bit of duplication in the job structure, but worth doing?

File: integration-tests/production-runtime/package.json [hunk 1]
```diff
@@ -5,10 +5,10 @@
   "author": "Kyle Mathews <mathews.kyle@gmail.com>",
   "dependencies": {
     "cypress": "^3.1.0",
-    "gatsby": "next",
-    "gatsby-plugin-manifest": "next",
-    "gatsby-plugin-offline": "next",
-    "gatsby-plugin-react-helmet": "next",
+    "gatsby": "latest",
```
Comment (DSchau) [inline integration-tests/production-runtime/package.json]: 👍 We should probably switch all versions in integration-tests to use this pattern now, too.

**Comment by KyleAMathews** [comment]
Maybe yeah. But since this is such an edge case (not many people are going to be PRing from older code), @pieh could just push a commit with a minor change to trigger tests

File: packages/gatsby/cache-dir/navigation.js [hunk 2]
```diff
@@ -139,4 +141,39 @@ function init() {
...
+    onRouteUpdate(this.props.location)
+  }
+
+  componentDidUpdate(prevProps, prevState, shouldFireRouteUpdate) {
+    if (shouldFireRouteUpdate) {
+      onRouteUpdate(this.props.location)
+    }
+  }
+
+  getSnapshotBeforeUpdate(prevProps, prevState) {
```
Comment (DSchau) [inline packages/gatsby/cache-dir/navigation.js]: Such a nit, but you're not actually using prevState are you?

Comment (pieh) [reply packages/gatsby/cache-dir/navigation.js]: ah, yeah, will fix - it should also trigger actual tests, so that's good

**Comment by KyleAMathews** [comment]
>sh: 1: gatsby: Permission denied

Hmmm that's why CircleCi failed @DSchau

**Comment by DSchau** [comment]
@KyleAMathews I can SSH into the box, but 99% sure that's the `chmod` fix that enables this to work. Otherwise I _could_ run the commands as sudo, which may also fix it.

**Comment by KyleAMathews** [comment]
Ok interesting — yeah, this sounds like a circleci wrinkle
